src/base/tl/algorithm.h

```diff
@@ -79,38 +80,271 @@ R find_binary(R range, T value)
+				{
+					right--;
+					*eright = *eleft;
+					eright--;
+				}
+			}
+
+			*eleft = pivot;
+			start[i+1] = left+1;
+			end[i+1] = end[i];
+			end[i++] = left;
+
+			// make sure to handle the smaller partition first
+			if(end[i]-start[i] > end[i-1]-start[i-1])
+			{
+				swap(start[i], start[i-1]);
+				swap(end[i], end[i-1]);
+			}
+		}
+		else
+			i--;
+	}
+}
+
+template<class R>
+void sort_heap(R range)
 {
 	concept_index::check(range);
-}*/
+	concept_forwarditeration::check(range);
+	concept_backwarditeration::check(range);
+
+	// build heap
+	int size = range.size();
+	typename R::type *ipe; // for inner loop
+	typename R::type *front = &range.front();
+	typename R::type *back = &range.back();
+	typename R::type *pe;
+	typename R::type *re;
+	for(pe = front+(size/2-1); pe >= front; pe--)
+	{
+		ipe = pe;
+		while(1)
+		{
+			re = front+((ipe-front)*2+1);
+			if(re > back)
+				break;
+
+			if(re+1 <= back)
+			{
+				if(*re < *(re+1))
+					re++;
+			}
+
+			if(*re < *ipe)
+				break;
+
+			swap(*re, *ipe);
+			ipe = re;
+		}
+	}
+
+	// do the final sorting
+	typename R::type *ire; // for inner loop
+	for(re = front+size-1; re > front; re--)
+	{
+		pe = front;
+		swap(*front, *re);
+		while(1)
+		{
+			ire = front+((pe-front)*2+1);
+			if(ire >= re)
+				break;
 
+			if(ire+1 < re)
+			{
+				if(*ire < *(ire+1))
+					ire++;
+			}
+
+			if(*ire < *pe)
+				break;
+
+			swap(*ire, *pe);
+			pe = ire;
+		}
+	}
+}
+
+template<class R>
+void sort_merge(R range)
+{
+	concept_forwarditeration::check(range);
+	concept_backwarditeration::check(range);
+
+	array<typename R::type> temp;
+	int size = range.size();
+	temp.set_size(size/2);
+	typename array<typename R::type>::range rtemp = temp.all();
+	typename R::type *front = &range.front();
+	typename R::type *tfront = &rtemp.front();
+	typename R::type *back = &range.back();
+	for(int s = 1; s < size; s += s)
+	{
+		for(typename R::type *m = back-1-s; m >= front; m -= s+s)
+		{
+			typename R::type *j;
+			if(m-s+1 < front)
+				j = front;
+			else
+				j = m-s+1;
+			typename R::type *b = tfront;
+			while(j <= m)
+			{
+				*b = *j;
+				b++;
+				j++;
+			}
+
+			typename R::type *k;
+			if(m-s+1 < front)
+				k = front;
+			else
+				k = m-s+1;
+
+			b = tfront;
+			while(k < j && j <= m+s)
+			{
+				if(!(*j < *b))
+				{
+					*k = *b;
+					k++;
+					b++;
+				}
+				else
+				{
+					*k = *j;
+					j++;
+					k++;
+				}
+			}
+
+			while(k < j)
+			{
+				*k = *b;
+				b++;
+				k++;
+			}
+		}
+	}
+}
+*/
+template<class R>
+void sort_shell(R range)
+{
+	// Using pseudocode from https://en.wikipedia.org/wiki/Shellsort
+	concept_forwarditeration::check(range);
+	concept_backwarditeration::check(range);
+
+	// Sort an array a[0...n-1].
+	/* gaps = [...] */
+	const int gaps[] = {
+		1, 3, 7, 21, 48, 112,
+		336, 861, 1968, 4592, 13776,
+		33936, 86961, 198768, 463792, 1391376,
+		3402672, 8382192, 21479367, 49095696, 114556624,
+		343669872, 52913488, 2085837936};
```

Comment:
Where do these gaps come from? I only see the first few in the Wikipedia article.